### PR TITLE
context, storagedriver: trace function calls to Base storage driver

### PR DESCRIPTION
**Pull request by stevvooe**
This adds `WithTrace` function to the context package that allows users to trace a time span between a trace allocation and returned function call. The resulting context includes ids that will allow for future dapper-like analysis. Please see the godoc addition for details.

This supports tracing calls to backend storage drivers.

This also moves the instance id out of the app so that it is associated with an instantiation of the runtime. The instance id is stored on the background context. This allows allow contexts using the main background context to include an instance id for log messages. It also simplifies the application slightly

### CODE REVIEW

**Comment by dmp42** [comment]
ping @aluzzardi @icecrime @jfrazelle @mavenugo

Should we share that kind of stuff?

File: context/trace.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+// 		defer done("this will be the log message")
+// 		// ... function body ...
+// 	}
+//
+// If the function ran for roughly 1s, such a usage would emit a log message
+// as follows:
+//
+// 	INFO[0001] this will be the log message  trace.duration=1.004575763s trace.func=github.com/docker/distribution/context.traceOperation trace.id=<id> ...
+//
+// Notice that the funciton name is automatically resolved, along with the
```
Comment (RichardScothern) [inline context/trace.go]: speeling.

Comment (stevvooe) [inline context/trace.go]: I just think you don't know what a "funciton" is.

**Comment by dmcgowan** [comment]
Given there is nothing in between info and debug, I would say Info is appropriate for such "chatty" logs.

LGTM

**Comment by stevvooe** [comment]
@dmcgowan Well, I was debating whether debug or info was correct. We'll go with this for now and see if it is too much. Maybe we can get @Sirupsen to add a Trace level for this use.

**Comment by RichardScothern** [comment]
lgtm

**Pull request merged**